File: src/algorithms/arrays/find-all-duplicates/find-all-duplicates.test.ts
import { describe, it, expect } from "vitest";
import { findAllDuplicates } from "./sources/find-all-duplicates.ts?fn";

describe("findAllDuplicates", () => {
  it("finds duplicates in the default input", () => {
    const result = findAllDuplicates([4, 3, 2, 7, 8, 2, 3, 1]);
    /* 2 and 3 appear twice */
    expect(result.sort()).toEqual([2, 3]);
  });

  it("returns empty array when no duplicates exist", () => {
    const result = findAllDuplicates([1, 2, 3, 4, 5]);
    expect(result).toEqual([]);
  });

  it("finds a single duplicate", () => {
    const result = findAllDuplicates([1, 2, 3, 2]);
    expect(result).toEqual([2]);
  });

  it("finds multiple duplicates", () => {
    const result = findAllDuplicates([1, 1, 2, 2, 3, 3]);
    expect(result.sort()).toEqual([1, 2, 3]);
  });

  it("handles a single-element array with no duplicates", () => {
    const result = findAllDuplicates([1]);
    expect(result).toEqual([]);
  });

  it("handles an empty array", () => {
    const result = findAllDuplicates([]);
    expect(result).toEqual([]);
  });

  it("handles all elements appearing twice", () => {
    const result = findAllDuplicates([2, 1, 2, 1]);
    expect(result.sort()).toEqual([1, 2]);
  });

  it("finds duplicates when they appear at the end of the array", () => {
    const result = findAllDuplicates([1, 2, 3, 4, 5, 4]);
    expect(result).toEqual([4]);
  });

  it("does not mutate the caller-visible view of the input", () => {
    const original = [4, 3, 2, 7, 8, 2, 3, 1];
    const snapshot = [...original];
    findAllDuplicates(original);
    /* The function copies internally, so original should be unchanged */
    expect(original).toEqual(snapshot);
  });
});

File: src/algorithms/arrays/cyclic-sort/find-all-duplicates/FindAllDuplicates_test.cpp
#include "sources/FindAllDuplicates.cpp"
#include <cassert>
#include <vector>
#include <algorithm>
#include <iostream>

int main() {
    // Default input [4,3,2,7,8,2,3,1] -> [2,3]
    {
        std::vector<int> result = findAllDuplicates({4, 3, 2, 7, 8, 2, 3, 1});
        std::sort(result.begin(), result.end());
        assert(result == std::vector<int>({2, 3}));
    }

    // No duplicates [1,2,3,4,5] -> []
    assert(findAllDuplicates({1, 2, 3, 4, 5}).empty());

    // Single duplicate [1,2,3,2] -> [2]
    assert(findAllDuplicates({1, 2, 3, 2}) == std::vector<int>({2}));

    // Multiple duplicates [1,1,2,2,3,3] -> [1,2,3]
    {
        std::vector<int> result = findAllDuplicates({1, 1, 2, 2, 3, 3});
        std::sort(result.begin(), result.end());
        assert(result == std::vector<int>({1, 2, 3}));
    }

    // Empty array -> []
    assert(findAllDuplicates({}).empty());

    std::cout << "All tests passed!" << std::endl;
    return 0;
}
